### PR TITLE
feat(bitbucket): Implement PR cache for repositories

### LINKED ISSUE #26278: Bitbucket Cloud: reduce and cache the PR lists
### Discussed in https://github.com/renovatebot/renovate/discussions/26275

<div type='discussions-op-text'>

<sup>Originally posted by **monwolf** December 13, 2023</sup>
### Tell us more.

I've been using renovatebot on a project for a while, this project has at least 1500 PRs closed and merged, and every time renovatebot tries to scan the project it takes more than 2 creating the PR list. 

To, improve the speed of scans on renovatebot, I would suggest using caches to save and reuse PR Lists and removing unused big fields from the request. 

To achieve this, Bitbucket Cloud provides two QS fields to filter and reduce the responses:

q can be used for adding a filter:

```
?q=updated_on > 2022-11-11T00:00:00-07:00
```

and fields to remove unused fields:

```
&fields=-values.description,-values.links,-values.summary
```

</div>

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Cache pull requests using repository cache

## Context

- Closes: #26278

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
Tested on real repository, looks okay

File: lib/modules/platform/bitbucket/index.ts [hunk 1]
```diff
@@ -894,8 +883,9 @@ export async function createPr({
         },
       )
     ).body;
+    BitbucketPrCache.init(config.repository, renovateUserUuid).addPr(prRes);
     const pr = utils.prInfo(prRes);
-    // istanbul ignore if
+    // istanbul ignore if: prList is initialized at this point
     if (config.prList) {
       config.prList.push(pr);
     }
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:891]:
```suggestion
     config.prList!.push(pr);
```
should we use this then?

File: lib/modules/platform/bitbucket/pr-cache.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  private getUrl(): string {
+    const params = new URLSearchParams();
+
+    for (const state of prStates.all) {
+      params.append('state', state);
+    }
+
+    params.append('fields', prFieldsFilter);
```
Comment (viceice) [inline lib/modules/platform/bitbucket/pr-cache.ts:74]: should we add an explicit order, so it never breaks in future?

**Comment by viceice** [comment]
can you implement the same for gitea please? 🤗😊

File: lib/modules/platform/bitbucket/index.ts (hunk 1 above)
Comment (zharinov) [reply lib/modules/platform/bitbucket/index.ts:891]: I would like to get rid of it, as it's most probably isn't a bottleneck

Comment (zharinov) [reply lib/modules/platform/bitbucket/index.ts:891]: But facing similar decision for GitHub, we left it intact

Comment (rarkins) [reply lib/modules/platform/bitbucket/index.ts]: Do we use it at runtime in GitHub, or is it designed to just make the cached list "less wrong" at the end?

File: lib/modules/platform/bitbucket/index.ts [hunk 3]
```diff
@@ -303,12 +288,13 @@ export async function findPr({
...
-      p.sourceBranch === branchName &&
-      (!prTitle || p.title.toUpperCase() === prTitle.toUpperCase()) &&
-      matchesState(p.state, state),
+  const pr = await getPrList().then((prs) =>
+    prs.find(
+      (p) =>
+        p.sourceBranch === branchName &&
+        (!prTitle || p.title.toUpperCase() === prTitle.toUpperCase()) &&
+        matchesState(p.state, state),
+    ),
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:297]: why this change?

File: lib/modules/platform/bitbucket/pr-cache.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    repoCache.platform.bitbucket.pullRequestsCache = pullRequestCache;
+    this.cache = pullRequestCache;
+  }
+
+  static async init(
+    http: BitbucketHttp,
+    repo: string,
+    author: string | null,
+  ): Promise<BitbucketPrCache> {
+    const prList = memCache.get<Pr[] | undefined>('bitbucket-pr-cache');
```
Comment (viceice) [inline lib/modules/platform/bitbucket/pr-cache.ts:41]: why checking memcache here, I don't see where it's written. I don't think we need that additional cache layer.

File: lib/modules/platform/bitbucket/index.ts (hunk 3 above)
Comment (zharinov) [reply lib/modules/platform/bitbucket/index.ts:297]: I don't know, intention just feels cleaner to me

File: lib/modules/platform/bitbucket/index.ts (hunk 1 above)
Comment (zharinov) [reply lib/modules/platform/bitbucket/index.ts]: Without something like `prList` (being stored ad-hoc or at the memory cache), it's harder to prevent `findPr()` from issuing at least one sync request on each call (it will call `getPrList()` under the hood).

Because of that, I came to conclusion we should keep the cache two-level, but incapsulate `prList` to the PR cache, instead of using it explicitly in the `index.ts` file.

File: lib/modules/platform/bitbucket/pr-cache.ts (hunk 2 above)
Comment (zharinov) [reply lib/modules/platform/bitbucket/pr-cache.ts:101]: Not sure what do you mean, I've added ordering below

Comment (zharinov) [reply lib/modules/platform/bitbucket/pr-cache.ts:101]: Ah, I see what you mean. Well, I'm not sure it's necessary since we're processing all the pages when uncached anyway, but once cached, we fetch only changed ones, so order doesn't matter so much, or is it?

File: lib/modules/platform/bitbucket/index.ts (hunk 3 above)
Comment (viceice) [reply lib/modules/platform/bitbucket/index.ts:315]: but with this change the code is more nested that before 🤔

I prefer await than then, especially when the callback function is more than two loc

File: lib/modules/platform/bitbucket/pr-cache.ts (hunk 2 above)
Comment (viceice) [reply lib/modules/platform/bitbucket/pr-cache.ts:101]: ok

File: lib/modules/platform/bitbucket/pr-cache.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  static async init(
+    http: BitbucketHttp,
+    repo: string,
+    author: string | null,
+  ): Promise<BitbucketPrCache> {
+    const prList = memCache.get<Pr[] | undefined>(
+      `bitbucket-pr-cache:${repo}:${author}`,
+    );
```
Comment (viceice) [inline lib/modules/platform/bitbucket/pr-cache.ts:43]: what's the benefit of using memcache over repo cache? both are memory cache, so this makes it more complicated.

the memory cache is always empty on repo start.

even if we do the http calls below again, they will returned from memcache. 🤔

Comment (zharinov) [reply lib/modules/platform/bitbucket/pr-cache.ts]: We have to store in memory cache at least the `isSynced` flag to avoid triggering query for every `findPr` call

**Review by viceice — APPROVED** [review]
